### ISSUE TITLE
Segmentation fault (core dumped)

### ISSUE THREAD

**Issue by gonzaparra** [s-u/rJava#110]
Hi all!

I am trying to reinstall a package that I was previously able to install and use. I was building a package of my own after my computer unexpectedly restarted and then I started to have problems loading the rpgraph package. So I decided to uninstall it and to reinstall it. When I did so I got the following error:

`library(devtools)
> library(rJava)
> install_github("Albluca/rpgraph")

Downloading GitHub repo Albluca/rpgraph@master
from URL https://api.github.com/repos/Albluca/rpgraph/zipball/master
Installing rpgraph
Running command /usr/lib/R/bin/R 
Arguments:
CMD
INSTALL
/tmp/Rtmp5OrtLL/devtools505a703b3ccd/Albluca-rpgraph-de04f96
--library=/home/gonzalo/R/x86_64-pc-linux-gnu-library/3.4
--install-tests

* installing *source* package ‘rpgraph’ ...
** R
** data
*** moving datasets to lazyload DB
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Segmentation fault (core dumped)
ERROR: loading failed
* removing ‘/home/gonzalo/R/x86_64-pc-linux-gnu-library/3.4/rpgraph’
Installation failed: run(bin, args = real_cmdargs, stdout_line_callback = real_callback(stdout),      stderr_line_callback = real_callback(stderr), stdout_callback = real_block_callback,      stderr_callback = real_block_callback, echo_cmd = echo, echo = show,      spinner = spinner, error_on_status = fail_on_status, timeout = timeout) : System command error`

I tried reinstalling JAva and cran from zero but somehow there is something that remains in my system that avoids the package to be installed. Also, since the error is not very informative, I have no idea on how to narrow where the problem is.

Thanks for any help you can provide.

**Comment by colearendt** [comment]
Having the same issue in the `xlsx` package.  Fairly confident it is related to security updates that were released for Ubuntu in the last few days (6/19, I think).  I fortunately had not installed the updates on my computer yet when the [issue came in](https://github.com/dragua/xlsx/issues/83), so I was able to confirm that installing the updates is what broke the package install.

Further, I have confirmed that the `.jpackage` function is what is causing the segfault.  It happens whether or not I use `.jpackage('xlsx')` or `.jpackage('xlsxjars')`, which have different jars included (so I do not suspect the Java code in particular).  To make matters more interesting, the segfault only seems to occur when `.jpackage` is called during `.onLoad` or `.onAttach`, and specifically when the package is being installed - calling it manually works after the package is installed.

I would suggest that this is a fairly high priority bug, as I would expect it to affect all Ubuntu OS's and all packages that use `rJava` with `jpackage` called in `.onLoad` or `.onAttach`.

**Comment by cs-dieter-kling** [comment]
>I would suggest that this is a fairly high priority bug, as I would expect it to affect all Ubuntu OS's and all packages that use rJava with jpackage called in .onLoad or >.onAttach.

*Absolutely*: `Segmentation fault` bug affects also `RJDBC` - we cannot access Hive DB any more after latest OS updates.
It tested with latest Debian, Centos, Ubuntu OSes and JDK 8 (Oracle JDK and OpenJDK) the error persists. 
A _workaround_ seems to be using older JDK version `1.7.0_95` but I'm still testing...

**Comment by colearendt** [comment]
Workaround used for the `xlsx` package was to temporarily suspend calling `.jpackage` from `.onLoad` / `.onAttach` when `grepl('linux-gnu',R.version$os)`.  This is less than ideal because it requires the user then to execute `.jpackage('xlsx')` manually - but hopefully not too much to ask if they are just now installing the package.

An avenue for exploring the root cause might be to pull apart the `.jpackage` function and execute line by line within `.onLoad` / `.onAttach`

**Comment by colearendt** [comment]
Very interesting.  Looks like the seg fault comes from `.jinit()` within installing a package for some reason - I do not know enough about R / Java internals to tell why - in particular, this piece of the `.jpackages` function:
```r
  if (!.jniInitialized) 
    .jinit()
```

Calling `.jinit()` from an interactive session has no issues.

**Comment by colearendt** [comment]
In particular, looks to be this line of `.jinit()`:
```r
xr <- .External(RinitJVM, boot.classpath, parameters)
```
Will need someone who is more versed than I in `C`/`C++` to dig into `RinitJVM` to see what the issue is.  Not sure how to nicely reproduce during install or if there is a way to diagnose by executing partial files / etc..  For testing, I have been building binary files locally (which will reproduce the segfault on Ubuntu) by running successive lines of code within `.onAttach`.

**Comment by gonzaparra** [comment]
mmm interesting but I have not updated any package in my OS so how can it be related to it? 

But yeah something weird is going on, since I am not really able to solve it. It works on other computers though so I guess that if I cannot solve it I will just need to make an OS reinstallation without updates maybe to be safe.

**Comment by colearendt** [comment]
@gonzaparra i would recommend forking the repo and removing any line like .jinit or .jpackage from the .onAttach or .onLoad functions and see if that allows you to install.  You should be able to call the functions yourself after installation and loading.

The restart you did probably allowed the problematic security updates to take effect

**Comment by gonzaparra** [comment]
Thanks for your answers.

I already reinstalled my OS. I installed the latest Ubuntu LTS version and
now everything works.. I hope these updated will be fixed cause is very
problematic.

Thanks for your time and answers if it happens again I hope I will be able
to solve it without formatting my pc.

Best
Gonza

**Comment by colearendt** [comment]
@s-u Not sure if you have seen, but this is a crippling bug for `rJava` on UNIX platforms related to a kernel change related to heap memory in the past week.  Best I can tell, the problem comes from 
```r
xr <- .External(RinitJVM, boot.classpath, parameters)
```

Update from the `xlsx` package: my workaround allows `install / library / .jpackage or .jinit`, but then `close R session / open new R session / library / .jpackage or .jinit` fails.

**Comment by cs-dieter-kling** [comment]
>@colearendt @s-u Not sure if you have seen, but this is a crippling bug for rJava on UNIX >platforms related to a kernel change related to heap memory in the past week.
I can confirm, in out case **RJDBC** fails with `Segmentation Fault` after security updtates of the Linux Kernel  
Old version:
`Linux 3.16.0-4-amd64 #1 SMP Debian 3.16.39-1 (2016-12-30) x86_64 GNU/Linux`
**new Kernel version**:
`Linux 3.16.0-4-amd64 #1 SMP Debian 3.16.43-2+deb8u1 (2017-06-18) x86_64 GNU/Linux`

lsb_release -a:
```
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 8.8 (jessie)
Release:    8.8
Codename:   jessie
```

**Comment by cs-dieter-kling** [comment]
Increasing stack size of the JVM by setting `-Xss2560k` Java option did the trick for me 
in order to get `RJDBC` to work:

```
options( java.parameters = c("-Xss2560k", "-Xmx2g") )
...

library(rJava)
library(RJDBC)

...

.jinit(classpath=c(hive.jar.path, hadoop.jar.path))

...
```

There are couple of workarounds have been suggested in order to counteract Segmentation faults:
* the above mentioned workaround - https://www.kvraudio.com/forum/viewtopic.php?p=6813554
* disabling _Address Space Layout Randomization_ - https://github.com/rust-lang/cargo/issues/4197 , https://askubuntu.com/questions/318315/how-can-i-temporarily-disable-aslr-address-space-layout-randomization

But it would be really interesting to understand what causes these issues and how to fix the root cause.

**Comment by colearendt** [comment]
No luck increasing stack size for the `xlsx` package, at least not for me - tried up to 10240k.  Other workarounds were a no-go as well, unfortunately.

I definitely agree that root cause would be great to know, as such debilitating issues are unsettling.

**Comment by gonzaparra** [comment]
Guys,

It seems that the ubuntu people have fixed the problem in today's kernel update release. 

I have updated my kernel to the  **v4.12-rc7/	2017-06-26 04:00**  version following the following steps suggested by WinEunuuchs2Unix on this thread https://askubuntu.com/questions/119080/how-to-update-kernel-to-the-latest-mainline-version-without-any-distro-upgrade

I am trying and so far I was able to use the rpgraph package again.

Best,

Gonzalo

**Comment by PRijnbeek** [comment]
We can confirm to experience the same behaviour when updating to kernel 4.4.0-81-generic. Rolling back to the previous version fixed the issue.

**Comment by jonassundman** [comment]
On ubuntu, running kernel version 4.4.0-81-generic, using rJava in a RStudio session still works as expected.

**Comment by DaFont** [comment]
On Ubuntu 4.8.0-58.63~16.04.1-generic 4.8.17, rJava works great again.

**Comment by colearendt** [comment]
For Ubuntu 14.04, kernel version 4.4.0-83-generic, rJava is working again as expected (it failed for me on 4.4.0-81-generic).  Looks like a kernel fix solved the problem.

**Comment by balwierz** [comment]
on Debian 9 running Linux 4.9.30 I am getting segfaults

**Comment by HappyRobotProject** [comment]
On Ubuntu  4.12.3-041203-generic  is not working

**Comment by colearendt** [comment]
The best test to see if you are having the same memory allocation on JVM startup issue is to load rJava and execute .jinit .  The issue is upstream from xlsx, as the JVM does not seem to start properly. Multiple potential fixes and workarounds discussed above

**Comment by rdauria** [comment]
I am having the same problem on CentOS 6.9 with 2.6.32-696.3.2.el6.x86_64 kernel. Any suggestion? Thanks.

**Comment by Jan-E** [comment]
@rdauria Try the CentOS 6.9 2.6.32-696.10.3.el6.x86_64 kernel (Oct 5)

**Comment by btreut** [comment]
I get the Segmentation fault with a fresh Siduction install (buster/sid) with kernel 4.14-6 (2017-11-30) during a call to JGR(). It looks like it is coming from JVM_GetStackAccessControlContext with traceback:
```
Java.lang.NullPointerException
	at org.rosuda.JGR.RController.getJgrDefaultPackages(RController.java:110)
	at org.rosuda.JGR.JGR.<init>(JGR.java:175)
	at org.rosuda.JGR.JGR.main(JGR.java:649)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at RJavaClassLoader.bootClass(RJavaClassLoader.java:532)
	at RJavaClassLoader.main(RJavaClassLoader.java:602)
```
OpenJDK 1.8.0_151

**Comment by markush81** [comment]
Above stacktrace is a follow up error because Rengine does not give any result. So not the cause od the issue itself.

**Comment by SolaimanR** [comment]
Hey. 

I think I have a similar issue, however, unrelated to rJava? The problems started when we did a restore from a backup, I believe a few permissions got mixed up and now R will give me the `Segmentation fault (core dumped)` error. The only way to start R is by typing `sudo R`. That works. Does anyone know how to fix it? I tried to change permissions in `/usr/lib/R `to my User using chown but it doesnt work.
